src/main/java/net/neoforged/neoforge/server/command/DataComponentCommand.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.server.command;
+
+import com.mojang.brigadier.Command;
+import com.mojang.brigadier.builder.ArgumentBuilder;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import java.util.Objects;
+import java.util.Optional;
+import net.minecraft.ChatFormatting;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.core.component.DataComponentMap;
+import net.minecraft.core.component.DataComponentPatch;
+import net.minecraft.core.component.DataComponentType;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.item.ItemStack;
+
+class DataComponentCommand {
+    public static ArgumentBuilder<CommandSourceStack, ?> register() {
+        return Commands.literal("data_components")
+                .then(Commands.literal("list")
+                        .requires(cs -> cs.hasPermission(Commands.LEVEL_GAMEMASTERS))
```

Comment:
Prefer putting the `hasPermission` requirement on the `data_components` node, so a player without the required permission won't be left with being able to see `/neoforge data_components` but nothing else down. (With my suggested change, a player without permission wouldn't even see `/neoforge data_components`.)